doc/sendmail.1
.TH sendmail 1
.SH NAME
sendmail \- sendmail emulator interface for nullmailer
.SH SYNOPSIS
.B sendmail
[
.B flags
] [
.I recipients
] <
.I message
.SH DESCRIPTION
This program is a front end program for
.B nullmailer-inject
and
.BR nullmailer-smtpd .
It is used by programs that expect a
.I sendmail
interface for sending email.
After parsing the command-line arguments, this program executes either
.B nullmailer-inject
or
.B nullmailer-smtpd
depending on the presence of the
.I \-bs
option on the command-line.
See the documentation for
.B nullmailer-inject
for details on how messages are reformatted and queued.
.SH OPTIONS
.TP
.B \-B TYPE
.TP
.B \-C FILE
.TP
.B \-d DEBUG
.TP
.B \-h COUNT
.TP
.B \-i
.TP
.B \-L TAG
.TP
.B \-N DSN
.TP
.B \-n
.TP
.B \-O OPTION
.TP
.B \-o OPTION
.TP
.B \-p PROTOCOL
.TP
.B \-q TIME
.TP
.B \-R RETURN
.TP
.B \-U
.TP
.B \-V ENVID
.TP
.B \-v
.TP
.B \-X LOGFILE
Ignored for compatibility
.TP
.B \-bm
Read mail from standard input (default).
.TP
.B \-bp
List information about mail queue. This executes
.BR mailq .
.TP
.B \-bs
Use the SMTP protocol on standard input and standard output. This
executes
.BR nullmailer-smtpd .
.TP
.B \-F ADDRESS
Sets the full name of the sender.
.TP
.B \-f ADDRESS
Sets the envelope sender address.
.TP
.B \-r ADDRESS
An alternate and obsolete form of the \-f flag.
.TP
.B \-t
Read message for recipients and ignore the command-line arguments.
.SH SEE ALSO
mailq(1),
nullmailer-inject(1),
nullmailer-smtpd(1)


NEWS
This file lists all the major user-visible changes to nullmailer.
-------------------------------------------------------------------------------
Changes in version 2.1

- Added support for TLS anonymous authentication.
  Thanks Uffe Jakobsen.

- Fixed sendmail wrapper handling of empty sender on command line.
  Thanks Sebastian Wiedenroth.

- Fixed handling of quoted strings in the "remotes" file.
  Thanks Mihai Moldovan.

- Fixed nullmailer-inject handling of leading "From " lines.

- Some build fixes.

- Fixed bogus temporary gethostbyname error message when the protocol
  source address was incorrect.

- Fixed potential race condition in tests.
  Thanks Felix Lechner.

- Fixed handling of time values on 32-bit big-endian systems.
  Thanks Felix Lechner.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 2.0

- Added support to nullmailer-send to move permanently failing messages
  out of the queue, and to generate bounce messages.

- Added support for IPv6.

- Added program to generate bounce/delay messages.

- Added an "allmailfrom" control file to nullmailer-queue, causing all
  messages to share a hard-coded envelope sender.

- Added logging the message sender/recipient in nullmailer-send.

- Improved handling of system errors when reading config files.

- Secured handling of password options for protocol modules.

- Support standard shell quoting for options in the "remotes" file.

- Added protocol option to set a separate TLS client private key file.

- Added protocol option to bind the source address on connections.

- Fixed nullmailer-inject to report errors to stderr.

- Fixed gnutls cast to pointer from integer of different size warning.

- Fixed nullmailer-inject and -queue to handle the null (empty) sender
  address. Needed for RFC 3798 (Message Disposition Notification).

- Moved spool directory to /var/spool/nullmailer like other MTAs.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.13

- Fixed growing recipient list when using sendmail -bs to send multiple
  messages.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.12

- Added a quadratic backoff for resend times.

- Added support for multiple admin addresses in nullmailer-queue.

- SMTP sender now sends a QUIT on success too.

- Fixed sendmail -bs mode to properly strip leading periods.

- Fixed duplicated final header line when message has no body.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.11

- Simplify (and expand) handling of ignored sendmail options.

- Fixed command-line interpretation of short options with values.

- Added support for SSL/TLS client certificates.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.10

- Added support for SMTPS (SMTP over SSL/TLS) and STARTTLS.

- Added support for auto-detection of login method in SMTP protocol.

- Added support for -bp (mailq) and -bs (SMTP) mode in the sendmail wrapper.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.06

- Made nullmailer-queue sync and not just flush queued files to protect
  against crashes during sensitive operations.

- Fixed handling of whitespace in quoted strings in headers.

- Fixed handling of timeouts when sending.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.05

- Fixed compilation of src/selfpipe.cc on Solaris and other OSes.

- Fixed handling of TM_HAS_ISDST on Solaris.

- Fixed a few address parsing glitches.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.04

- Fixed rewriting mail to @localhost if defaultdomain is set by not
  appending .defaultdomain to *@localhost.

- Added support for reporting the sender and recipients in mailq.

- Added a send timeout in nullmailer-send, to kill sending messages that
  stall.

- Added support in nullmailer-inject for $NULLMAILER_QUEUE to name the
  program which is used to queue messages.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.03

- Fixed the SMTP protocol module to send QUIT on protocol failures.

- Added support for the SMTP AUTH LOGIN method.  Thanks to Sean
  MacLennan for the suggestion.

- Fixed the SMTP protocol module to send EHLO instead of HELO when doing
  authentication.  Thanks to Norbert Tretkowski for pointing this out.

- Fixed compile breakage on Solaris due to use of setenv.

- Fixed netstring length bug in QMQP sending module.

- If the sender name cannot be determined from the environment
  variables, try to pull it from /etc/passwd before using "unknown".
  Thanks Roderick Schertler.


src/sendmail.cc
// nullmailer -- a simple relay-only MTA
// Copyright (C) 2017  Bruce Guenter <bruce@untroubled.org>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
// You can contact me at <bruce@untroubled.org>.  There is also a mailing list
// available to discuss this package.  To subscribe, send an email to
// <nullmailer-subscribe@lists.untroubled.org>.

#include "config.h"
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "configio.h"
#include "fdbuf/fdbuf.h"
#include "defines.h"
#include "forkexec.h"
#include "setenv.h"
#include "cli++/cli++.h"

const char* cli_program = "sendmail";
const char* cli_help_prefix = "Nullmailer sendmail emulator\n";
const char* cli_help_suffix = 0;
const char* cli_args_usage = "[recipients] <message";
const int cli_args_min = 0;
const int cli_args_max = -1;
const bool cli_only_long = true;

enum mode { mode_normal, mode_mailq, mode_smtp };

static int o_dummyi;
static const char* o_dummys;
static const char* o_sender = 0;
static int o_mode = 0;
static char* o_from;
static int use_header = false;

cli_option cli_options[] = {
  { 'B', 0,    cli_option::string, 0, &o_dummys, "Ignored", 0 },
  {  0, "bm",  cli_option::flag, mode_normal, &o_mode,
     "Read mail from standard input (default)", 0 },
  {  0, "bp",  cli_option::flag, mode_mailq, &o_mode,
     "List information about mail queue", 0 },
  {  0, "bs",  cli_option::flag, mode_smtp, &o_mode,
     "Handle SMTP commands on standard input", 0 },
  { 'C', 0,    cli_option::string, 0, &o_dummys, "Ignored", 0 },
  { 'd', 0,    cli_option::string, 0, &o_dummys, "Ignored", 0 },
  { 'F', 0,    cli_option::string, 0, &o_sender,
    "Set the full name of the sender", 0 },
  { 'f', 0,    cli_option::string, 0, &o_from,
    "Set the envelope sender address", 0 },
  { 'h', 0,    cli_option::string, 0, &o_dummys, "Ignored", 0 },
  { 'i', 0,    cli_option::flag, 0, &o_dummyi, "Ignored", 0 },
  { 'L', 0,    cli_option::string, 0, &o_dummys, "Ignored", 0 },
  { 'm', 0,    cli_option::flag,   0, &o_dummyi, "Ignored", 0 },
  { 'N', 0,    cli_option::string, 0, &o_dummys, "Ignored", 0 },
  { 'n', 0,    cli_option::flag, 0, &o_dummyi, "Ignored", 0 },
  { 'O', 0,    cli_option::string, 0, &o_dummys, "Ignored", 0 },
  { 'o', 0,    cli_option::string, 0, &o_dummys, "Set sendmail option, ignored", 0 },
  {  0, "em",  cli_option::flag, 0, &o_dummyi,
     "Ignored", 0 },
  {  0, "ep",  cli_option::flag, 0, &o_dummyi,
     "Ignored", 0 },
  {  0, "eq",  cli_option::flag, 0, &o_dummyi,
     "Ignored", 0 },
  { 'p', 0,    cli_option::string, 0, &o_dummys, "Ignored", 0 },
  { 'q', 0,    cli_option::string, 0, &o_dummys, "Ignored", 0 },
  { 'R', 0,    cli_option::string, 0, &o_dummys, "Ignored", 0 },
  { 'r', 0,    cli_option::string, 0, &o_from,
    "An alternate and obsolete form of the -f flag", 0 },
  { 't', 0,    cli_option::flag, 1, &use_header,
    "Read message for recipients", 0 },
  { 'U', 0,    cli_option::flag, 0, &o_dummyi, "Ignored", 0 },
  { 'V', 0,    cli_option::string, 0, &o_dummys, "Ignored", 0 },
  { 'v', 0,    cli_option::flag, 0, &o_dummyi, "Ignored", 0 },
  { 'X', 0,    cli_option::string, 0, &o_dummys, "Ignored", 0 },
  CLI_OPTION_END
};

bool setenvelope(char* str)
{
  char* at = strchr(str, '@');
  if(at) {
    *at = 0;
    setenv("NULLMAILER_HOST", at+1, 1);
  }
  setenv("NULLMAILER_USER", str, 1);
  return true;
}

int do_exec(const char* program, const char* args[])
{
  mystring path = program_path(program);
  args[0] = path.c_str();
  execv(args[0], (char**)args);
  ferr << "sendmail: Could not exec " << program << '.' << endl;
  return 1;
}

int cli_main(int argc, char* argv[])
{
  const char* extra_args[argc + 5] = {0};
  int extra_argc = 1;

  switch (o_mode) {
  case mode_smtp:
    return do_exec("nullmailer-smtpd", extra_args);
  case mode_mailq:
    return do_exec("mailq", extra_args);
  default:

    extra_args[extra_argc++] = use_header ? "-b" : "-e";

    if(o_sender)
      setenv("NULLMAILER_NAME", o_sender, 1);
    if(o_from) {
      if(!setenvelope(o_from))
        return -1;

      // pass along empty sender
      if(o_from[0] == '\0' || (strcmp(o_from, "<>") == 0)) {
        extra_args[extra_argc++] = "-f";
        extra_args[extra_argc++] = o_from;
      }
    }

    for(int i = 0; i < argc; i++)
      extra_args[extra_argc++] = argv[i];

    return do_exec("nullmailer-inject", extra_args);
  }
}
